nexus/db-queries/src/db/queries/sled_reservation.rs

```diff
@@ -347,11 +392,138 @@ pub fn sled_insert_resource_query(
         ),",
     );
 
+    match local_storage_allocation_required {
+        LocalStorageAllocationRequired::No => {}
+
+        LocalStorageAllocationRequired::Yes { allocations } => {
+            // Update each local storage disk's allocation id
+
+            query.sql(" UPDATED_LOCAL_STORAGE_DISK_RECORDS AS (");
+            query.sql("  UPDATE disk_type_local_storage ");
+
+            query.sql(
+                "    SET local_storage_dataset_allocation_id = CASE disk_id ",
+            );
+
+            for allocation in allocations {
+                let LocalStorageAllocation {
+                    disk_id,
+                    local_storage_dataset_allocation_id,
+                    ..
+                } = allocation;
+
+                query.sql("WHEN ");
+                query.param().bind::<sql_types::Uuid, _>(*disk_id);
+                query.sql(" THEN ");
+                query.param().bind::<sql_types::Uuid, _>(
+                    local_storage_dataset_allocation_id.into_untyped_uuid(),
+                );
+                query.sql(" ");
+            }
+
+            query.sql("    END");
+
+            query.sql("  WHERE disk_id in ( ");
+
+            for (index, allocation) in allocations.iter().enumerate() {
+                let LocalStorageAllocation { disk_id, .. } = allocation;
+
+                query.param().bind::<sql_types::Uuid, _>(*disk_id);
+
+                if index != (allocations.len() - 1) {
+                    query.sql(",");
+                }
+            }
+
```

Comment:
there's enough Rust control flow intermingled with the query fragments here that i think some comments showing what we're trying to construct here could be worth having --- i know there's an expectorate test, but i think it could still be nice.